libraries/AP_FETtecOneWire/AP_FETtecOneWire.cpp

```diff
@@ -0,0 +1,732 @@
+        update_rpm(tlm_from_id-1, centi_erpm*100*2/pole_count.get(), tx_err_rate);
+
+        update_telem_data(tlm_from_id-1, t, AP_ESC_Telem_Backend::TelemetryType::TEMPERATURE|AP_ESC_Telem_Backend::TelemetryType::VOLTAGE|AP_ESC_Telem_Backend::TelemetryType::CURRENT|AP_ESC_Telem_Backend::TelemetryType::CONSUMPTION);
+    }
+#endif
+
+}
+
+/**
+    generates used 8 bit CRC for arrays
+    @param buf 8 bit byte array
+    @param buf_len count of bytes that should be used for CRC calculation
+    @return 8 bit CRC
+*/
+uint8_t AP_FETtecOneWire::get_crc8(uint8_t* buf, uint16_t buf_len) const
+{
+    uint8_t crc = 0;
+    for (uint16_t i = 0; i < buf_len; i++) {
+        crc = crc8_dvb(buf[i], crc, 0x7);
+    }
+    return (crc);
+}
+
+/**
+    transmits a FETtec OneWire frame to an ESC
+    @param esc_id id of the ESC
+    @param bytes 8 bit array of bytes. Where byte 1 contains the command, and all following bytes can be the payload
+    @param length length of the bytes array
+*/
+void AP_FETtecOneWire::transmit(uint8_t esc_id, uint8_t* bytes, uint8_t length)
+{
+    /*
+    a frame looks like:
+    byte 1 = frame header (master is always 0x01)
+    byte 2 = target ID (5bit)
+    byte 3 & 4 = frame type (always 0x00, 0x00 used for bootloader. here just for compatibility)
+    byte 5 = frame length over all bytes
+    byte 6 - X = request type, followed by the payload
+    byte X+1 = 8bit CRC
+    */
+    uint8_t transmit_arr[256] = {0x01, esc_id, 0x00, 0x00};
+    transmit_arr[4] = length + 6;
+    for (uint8_t i = 0; i < length; i++) {
+        transmit_arr[i + 5] = bytes[i];
+    }
+    transmit_arr[length + 5] = get_crc8(transmit_arr, length + 5); // crc
+    _uart->write_locked(transmit_arr, length + 6,FTOW_UART_LOCK_KEY);
+}
+
+/**
+    reads the FETtec OneWire answer frame of an ESC
+    @param bytes 8 bit byte array, where the received answer gets stored in
+    @param length the expected answer length
+    @param return_full_frame can be OW_RETURN_RESPONSE or OW_RETURN_FULL_FRAME
+    @return 2 on CRC error, 1 if the expected answer frame was there, 0 if dont
+*/
+uint8_t AP_FETtecOneWire::receive(uint8_t* bytes, uint8_t length, return_type return_full_frame)
+{
+    /*
+    a frame looks like:
+    byte 1 = frame header (0x02 = bootloader, 0x03 = ESC firmware)
+    byte 2 = sender ID (5bit)
+    byte 3 & 4 = frame type (always 0x00, 0x00 used for bootloader. here just for compatibility)
+    byte 5 = frame length over all bytes
+    byte 6 - X = answer type, followed by the payload
+    byte X+1 = 8bit CRC
+    */
+
+    // look for the real answerOW_SET_TLM_TYPE
+    if (_uart->available() >= length + 6u) {
+        // sync to frame starte byte
+        uint8_t test_frame_start = 0;
+        do {
+            test_frame_start = _uart->read();
+        }
+        while (test_frame_start != 0x02 && test_frame_start != 0x03 && _uart->available());

```

Comment:
better to use the time gap for rx, if first byte isn't 2 or 3, then call uart->discard_input() and throw it all away

Reply:
I thought about that and I do not know, if I got your point. If there are some bytes in front of the current package (which shouldn't be the normal case) they are dismissed until the start of newest package is found.
This is a safe way to find the startpoint. Doesn't it require some luck to get and stay in sync if I discard all the packages while the receive function and the bufferstuff execution is async?

You point is that it could stay in the loop, if something crazy spams on the line right? I could add an additional timeout to avoid that.

Reply:
I now call the uart->discard_input() method after reading the expected data. That uses the time gap to do some of the work for us.